core-services/nvidia_gpu.tf

```diff
@@ -0,0 +1,91 @@
+locals {
+  gpu_regex                  = "^(p[0-9][a-z]*|g[0-9+][a-z]*|trn[0-9][a-z]*|inf[0-9]|dl[0-9][a-z]*|f[0-9]|vt[0-9])\\..*"
+  has_gpu_node_groups        = length([for ng in var.node_groups : ng if length(regexall(local.gpu_regex, ng.instance_type)) > 0]) > 0
+  gpu_enabled_instance_types = [for ng in var.node_groups : ng.instance_type if length(regexall(local.gpu_regex, ng.instance_type)) > 0]
+}
+
+resource "kubernetes_daemonset" "nvidia" {
+  count = local.has_gpu_node_groups ? 1 : 0
+  metadata {
+    name      = "nvidia-device-plugin-daemonset"
+    namespace = kubernetes_namespace_v1.md-core-services.metadata.0.name
+    labels = merge(var.md_metadata.default_tags, {
+      k8s-app = "nvidia-device-plugin-daemonset"
+    })
+  }
+  spec {
+    selector {
+      match_labels = {
+        name = "nvidia-device-plugin-ds"
+      }
+    }
+    strategy {
+      type = "RollingUpdate"
+    }
+    template {
+      metadata {
+        labels = merge(var.md_metadata.default_tags, {
+          name = "nvidia-device-plugin-ds"
+        })
+        annotations = {
+          "scheduler.alpha.kubernetes.io/critical-pod" : ""
+        }
+      }
+      spec {
+        priority_class_name = "system-node-critical"
+        affinity {
+          node_affinity {
+            required_during_scheduling_ignored_during_execution {
+              node_selector_term {
+                match_expressions {
+                  key      = "node.kubernetes.io/instance-type"
+                  operator = "In"
+                  values   = local.gpu_enabled_instance_types
+                }
+              }
+            }
+          }
+        }
+        toleration {
+          key      = "CriticalAddonsOnly"
+          operator = "Exists"
+        }
+        toleration {
```

Comment:
Why is this toleration needed? We aren't applying it, are we?

Reply:
It comes with the NVIDIA daemonset plugin: https://raw.githubusercontent.com/NVIDIA/k8s-device-plugin/v0.14.0/nvidia-device-plugin.yml

And then used by GPU pods it seems? https://github.com/NVIDIA/k8s-device-plugin?tab=readme-ov-file#running-gpu-jobs

Reply:
I think they are just using that as the taint in place of the `gpu=true` taint we added. We probably don't need both. We can use theirs instead.